community/kernel/src/main/java/org/neo4j/kernel/impl/api/FreeIdManagement.java

```diff
@@ -0,0 +1,49 @@
+/*
+ * Copyright (c) 2002-2016 "Neo Technology,"
+ * Network Engine for Objects in Lund AB [http://neotechnology.com]
+ *
+ * This file is part of Neo4j.
+ *
+ * Neo4j is free software: you can redistribute it and/or modify
+ * it under the terms of the GNU General Public License as published by
+ * the Free Software Foundation, either version 3 of the License, or
+ * (at your option) any later version.
+ *
+ * This program is distributed in the hope that it will be useful,
+ * but WITHOUT ANY WARRANTY; without even the implied warranty of
+ * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+ * GNU General Public License for more details.
+ *
+ * You should have received a copy of the GNU General Public License
+ * along with this program.  If not, see <http://www.gnu.org/licenses/>.
+ */
+package org.neo4j.kernel.impl.api;
+
+import org.neo4j.kernel.impl.store.CommonAbstractStore;
+
+/**
+ * Responsible for freeing ids of deleted records. Deleting records does not always mean that that
+ * id will be free for use there and then. Transaction control and isolation semantics comes into play,
+ * that is why this aspect is pulled out from the stores and managed here.
+ */
+public interface FreeIdManagement
```

Comment:
I find this abstraction weird. It's asymmetrical and I don't see why it's in `kernel.impl.api`. Could we decorate the `IdGenerator` in the stores instead, and let the decorating `IdGeneratorFactory` expose a thing to control its internal state wrt. recovery and the periodic maintenance runs?